Backend/Remora.Discord.Caching/Extensions/CacheBuilder.cs

```diff
@@ -0,0 +1,46 @@
+//
+//  CacheBuilder.cs
+//
+//  Author:
+//       Jarl Gullberg <jarl.gullberg@gmail.com>
+//
+//  Copyright (c) 2017 Jarl Gullberg
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Lesser General Public License for more details.
+//
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Remora.Discord.Caching.Extensions
+{
+    /// <summary>
+    /// Utility class used by <see cref="ServiceCollectionExtensions.AddDiscordCaching"/> to configure the cache.
+    /// </summary>
+    public class CacheBuilder
```

Comment:
Any particular reason for this type? The one place it's being used could be replaced with an `Action<IServiceCollection>`, as far as I can tell - either that, or omitting it completely, simply letting the user call `UseMemory` or `UseRedis`, and altering those extension methods to add caching on their own. That way, you also avoid a user accidentally not registering any caching backend (if that route is taken, the methods should probably be renamed, too).

Reply:
One option is also to fall back to memory caching if no other strategy has been configured. Perhaps an argument for retaining the memory cache in the main Caching package?

Reply:
> Any particular reason for this type? The one place it's being used could be replaced with an `Action<IServiceCollection>`

The only downside is that you'll get all the extension methods of the interface `IServiceCollection`. For example you could do the following:
```cs
services.AddDiscordCaching(x => x.AddDiscordCaching(y => y.AddDiscordCaching())); // just to make sure it gets registered 😄
```

With this new type, you'll only get the extensions methods that targets the cache building.
```cs
services.AddDiscordCaching(b => b.UseRedis());
```

> simply letting the user call UseMemory or UseRedis, and altering those extension methods to add caching on their own

From my understanding, you want to have a new extensions method called for example `services.AddDiscordRedisCaching()`?

> One option is also to fall back to memory caching if no other strategy has been configured.

I've added this. When no cache client is registered, the memory cache client will be added.

Noteworthy is that the Memory Cache will always be registered (even when Redis is being used) but maybe this is for the best, otherwise it would be a breaking change (if users depend on this without calling AddMemoryCache because AddDiscordCaching already does this).

Reply:
Alright - looks good now :)